File: go/apps_test.go
package buble

import (
	"context"
	"encoding/json"
	"net/http"
	"testing"
	"time"
)

func TestAppsListAndRetrieve(t *testing.T) {
	client, server := newTestClient(func(w http.ResponseWriter, r *http.Request) {
		switch r.URL.Path {
		case "/api/v1/apps":
			if r.URL.Query().Get("limit") != "10" {
				t.Fatalf("missing limit query: %s", r.URL.RawQuery)
			}
			writeJSON(t, w, http.StatusOK, map[string]any{"data": []any{map[string]any{"id": "video-background-remover", "input_parameters": []any{}}}})
		case "/api/v1/apps/video-background-remover":
			writeJSON(t, w, http.StatusOK, map[string]any{"data": map[string]any{"id": "video-background-remover", "input_parameters": []any{}}})
		default:
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
	})
	defer server.Close()

	apps, err := client.Apps.List(context.Background(), 0, 10)
	if err != nil {
		t.Fatalf("Apps.List returned error: %v", err)
	}
	if apps.Data[0].ID != "video-background-remover" {
		t.Fatalf("unexpected apps: %#v", apps.Data)
	}
	app, err := client.Apps.Retrieve(context.Background(), "video-background-remover")
	if err != nil {
		t.Fatalf("Apps.Retrieve returned error: %v", err)
	}
	if app.Data.ID != "video-background-remover" {
		t.Fatalf("unexpected app: %#v", app.Data)
	}
}

func TestAppGenerationCreateAndWait(t *testing.T) {
	statuses := []TaskStatus{StatusPending, StatusSuccess}
	var body map[string]any
	client, server := newTestClient(func(w http.ResponseWriter, r *http.Request) {
		switch r.Method + " " + r.URL.Path {
		case "POST /api/v1/apps/video-background-remover/generations":
			if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
				t.Fatalf("decode request: %v", err)
			}
			writeJSON(t, w, http.StatusCreated, map[string]any{"data": map[string]any{"id": "app_task_1", "status": "pending"}})
		case "GET /api/v1/apps/video-background-remover/generations/app_task_1":
			status := statuses[0]
			statuses = statuses[1:]
			writeJSON(t, w, http.StatusOK, map[string]any{"data": map[string]any{"id": "app_task_1", "status": status}})
		default:
			t.Fatalf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
	})
	defer server.Close()

	task, err := client.Apps.Generations.Create(context.Background(), "video-background-remover", map[string]any{
		"source_video": []string{"https://example.com/source.mp4"},
	})
	if err != nil {
		t.Fatalf("Create returned error: %v", err)
	}
	if body["source_video"] == nil || task.Data.ID != "app_task_1" {
		t.Fatalf("unexpected app generation body/task: %#v %#v", body, task.Data)
	}
	result, err := client.Apps.Generations.Wait(
		context.Background(),
		"video-background-remover",
		"app_task_1",
		WithWaitInterval(time.Millisecond),
		WithWaitTimeout(time.Second),
	)
	if err != nil {
		t.Fatalf("Wait returned error: %v", err)
	}
	if result.Data.Status != StatusSuccess {
		t.Fatalf("unexpected status: %s", result.Data.Status)
	}
}

File: elixir/test/buble/apps_test.exs
defmodule Buble.AppsTest do
  use ExUnit.Case, async: true

  setup do
    {:ok, client} = Buble.Client.new(api_key: "sk-test", transport: Buble.TestTransport)
    %{client: client}
  end

  test "lists and retrieves apps", %{client: client} do
    Process.put(:buble_test_handler, fn _client, method, path, _opts ->
      assert method == :get
      assert path in ["/api/v1/apps", "/api/v1/apps/video-background-remover"]
      {:ok, %{"data" => []}}
    end)

    assert {:ok, _} = Buble.Apps.list(client)
    assert {:ok, _} = Buble.Apps.retrieve(client, "video-background-remover")
  end

  test "creates app generations", %{client: client} do
    Process.put(:buble_test_handler, fn _client, method, path, opts ->
      assert method == :post
      assert path == "/api/v1/apps/video-background-remover/generations"
      assert Keyword.fetch!(opts, :json) == %{"video_url" => "https://example.test/input.mp4"}
      {:ok, %{"data" => %{"id" => "app_gen_123"}}}
    end)

    assert {:ok, %{"data" => %{"id" => "app_gen_123"}}} =
             Buble.Apps.Generations.create(client, "video-background-remover",
               video_url: "https://example.test/input.mp4"
             )
  end
end
